keyvault/client/src/main/java/com/azure/keyvault/SecretClient.java

```diff
@@ -0,0 +1,525 @@
+        return getPagedSecrets(service.getSecretVersions(vaultEndPoint, name, maxPageResults, API_VERSION, ACCEPT_LANGUAGE));
+    }
+
+    /**
+     * Gets attributes of all the secrets given by the {@code nextPageLink} that was retrieved from a call to
+     * {@link SecretClient#listSecretsAsync()} or {@link SecretClient#listSecretsAsync(int)}.
+     *
+     * @param nextPageLink The {@link Page#nextPageLink()} from a previous, successful call to one of the list operations.
+     * @return A stream of {@link SecretAttributes} from the next page of results.
+     */
+    private Flux<SecretAttributes> listSecretsNext(String nextPageLink) {
+        return getPagedSecrets(service.getSecrets(vaultEndPoint, nextPageLink, ACCEPT_LANGUAGE));
+    }
+
+    /**
+     * Gets all deleted secrets given by the {@code nextPageLink} that was retrieved from a call to
+     * {@link SecretClient#listDeletedSecretsAsync()} or {@link SecretClient#listDeletedSecretsAsync(int)}.
+     *
+     * @param nextPageLink The {@link Page#nextPageLink()} from a previous, successful call to one of the list operations.
+     * @return A {@link Flux} containing {@link SecretAttributes} of all the versions of the secret in the key vault.
+     * @return A stream of {@link DeletedSecret} from the next page of results.
+     */
+    private Flux<DeletedSecret> listDeletedSecretsNext(String nextPageLink) {
+        return getPagedDeletedSecrets(service.getDeletedSecrets(vaultEndPoint, nextPageLink, ACCEPT_LANGUAGE));
+    }
+
+    private Flux<SecretAttributes> getPagedSecrets(Mono<RestResponse<Page<SecretAttributes>>> response) {
+        return response.flatMapMany(p -> Flux.just(new RestPagedResponseImpl<>(p.body().items(), p.body().nextPageLink(), p.request(), p.headers(), p.statusCode())))
+                .concatMap(this::extractAndFetchSecrets);
+    }
+
+    private Flux<DeletedSecret> getPagedDeletedSecrets(Mono<RestResponse<Page<DeletedSecret>>> response) {
+        return response.flatMapMany(p -> Flux.just(new RestPagedResponseImpl<>(p.body().items(), p.body().nextPageLink(), p.request(), p.headers(), p.statusCode())))
+                .concatMap(this::extractAndFetchDeletedSecrets);
+    }
+
+    private Publisher<SecretAttributes> extractAndFetchSecrets(RestPagedResponseImpl<SecretAttributes> page) {
+        String nextPageLink = page.nextLink();
+        if (nextPageLink == null) {
+            return Flux.fromIterable(page.items());
+        }
+        return Flux.fromIterable(page.items()).concatWith(listSecretsNext(nextPageLink));
+    }
+
+    private Publisher<DeletedSecret> extractAndFetchDeletedSecrets(RestPagedResponseImpl<DeletedSecret> page) {
+        String nextPageLink = page.nextLink();
+        if (nextPageLink == null) {
+            return Flux.fromIterable(page.items());
+        }
+        return Flux.fromIterable(page.items()).concatWith(listDeletedSecretsNext(nextPageLink));
+    }
```

Comment:
I think these two methods could be extracted and made generic also. Ping me sometime and I'll share my thinking.

Reply:
Should we put one time effort in doing this in azure-common?

Reply:
If it can be made general I would do it. I suspect it can be made more general.

Reply:
I tried an approach using generic methods, not pushed to repo yet.
I will run it over with you on Teams call.